### PR TITLE
Charge transaction fee even in case of ProgramError

### LINKED ISSUE #2526: Fee is not collected on transactions that fail with a ProgramError
While we write transactions that fail with a ProgramError to the ledger, it appears that we don't actually collect the fee for such a transaction.   As the return is not `Ok()` from such transactions, the bank doesn't store the accounts.  This is correct but the fee is collected in the `Ok()` case as a side-effect of storing, so on a ProgramError we need to do a little additional work to still debt the fee.

### PR DESCRIPTION
**Pull request by pgarg66**
#### Problem
The fee is not charged when a transaction fails due to ProgramError

#### Summary of Changes
Check if the result is a ProgramError. If so, withdraw transaction fee from account 0.
 
Fixes #2526

### CODE REVIEW

**Comment by pgarg66** [comment]
I think a similar change is needed in BankingStage. But, I am a bit confused with the current code flow and architecture changes coming soon.

**Comment by garious** [comment]
I think we were looking for a way to remove the special case rather than add a new one. Any way to do that?

**Comment by pgarg66** [comment]
Special case of checking for ProgramError?

**Comment by rob-solana** [comment]
I'd expect the same code to appear in process_transactions() (i.e. on the TPU side) which would obviate the need for this code at the level of processing entries.  this code should be in runtime...

File: runtime/src/bank.rs [hunk 1]
```diff
@@ -519,6 +519,12 @@ impl Bank {
         parents
     }
 
+    pub fn withdraw(&self, pubkey: &Pubkey, tokens: u64) {
+        let mut account = self.get_account(pubkey).unwrap_or_default();
```
Comment (garious) [inline runtime/src/bank.rs]: If `None`, this should return an error.

**Comment by garious** [comment]
@pgarg66, close this one in favor of the `process_transactions()` approach?

**Comment by pgarg66** [comment]
> @pgarg66, close this one in favor of the `process_transactions()` approach?

@garious , I've updated this PR with the new approach.

File: runtime/src/bank.rs [hunk 2]
```diff
@@ -512,6 +517,27 @@ impl Bank {
             txs.len(),
         );
         self.update_transaction_statuses(txs, &executed);
+
+        let mut fees = 0;
```
Comment (garious) [inline runtime/src/bank.rs]: Can you put all of this into a new function? Looks like a lot of edge cases we'd want to test independently of the code around it.

Comment (garious) [reply runtime/src/bank.rs]: That can happen in a follow-up PR

**Comment by pgarg66** [comment]
This PR should also fix https://github.com/solana-labs/solana/issues/1157

File: runtime/src/bank.rs (hunk 2 above)
Comment (pgarg66) [reply runtime/src/bank.rs]: 👍 will do.

**Comment by pgarg66** [comment]
Fixes #1157

**Pull request merged**